src/httprpc.cpp

```diff
@@ -200,13 +203,17 @@ static bool HTTPReq_JSONRPC(const std::any& context, HTTPRequest* req)
                 req->WriteReply(HTTP_FORBIDDEN);
                 return false;
             }
-            UniValue result = tableRPC.execute(jreq);
 
-            // Send reply
-            strReply = JSONRPCReply(result, NullUniValue, jreq.id);
+            reply = JSONRPCExec(jreq);
+            if (jreq.IsNotification()) {
+                // Even though we do execute notifications, we do not respond to them
+                req->WriteReply(HTTP_NO_CONTENT);
+                return true;
+            }
 
         // array of requests
         } else if (valRequest.isArray()) {
```

Comment:
This is a good catch but unfortunately I think we need to stick with this behavior for backwards compatibility. If we get a request that does not have the exact key-value pair `"jsonrpc":"2.0"` we have to resort to legacy behavior. In that case, the user has submitted an empty batch and what they get back is an empty vector.

The closest 2.0 example I can think of is this:

```
--> [{"jsonrpc":"2.0", "id":""}]
<-- [{'jsonrpc': '2.0', 'error': {'code': -32600, 'message': 'Missing method'}, 'id': ''}]
```

if `"jsonrpc"` is removed, we revert to legacy behavior and if `"id"` is removed we get no response at all since that is a notification:

```
--> [{"jsonrpc":"2.0"}]
<-- []
```

Reply:
Thanks!!!

Fixed in force push to 771d1e1d206efe687b8661ab966cc1a62cc7ba39



```diff
diff --git a/src/httprpc.cpp b/src/httprpc.cpp
index 54354a8625..f678668fd3 100644
--- a/src/httprpc.cpp
+++ b/src/httprpc.cpp
@@ -247,6 +247,11 @@ static bool HTTPReq_JSONRPC(const std::any& context, HTTPRequest* req)
                 }
                 if (!jreq.IsNotification()) reply.push_back(std::move(response));
             }
+            if (reply.empty()) {
+                // All-notification batch expects no response
+                req->WriteReply(HTTP_NO_CONTENT);
+                return true;
+            }
         }
         else
             throw JSONRPCError(RPC_PARSE_ERROR, "Top-level object parse error");
diff --git a/test/functional/interface_rpc.py b/test/functional/interface_rpc.py
index 309345731c..dcb0b66921 100755
--- a/test/functional/interface_rpc.py
+++ b/test/functional/interface_rpc.py
@@ -161,6 +161,14 @@ class RPCInterfaceTest(BitcoinTestFramework):
             ]
         )
 
+        self.log.info("Testing JSON-RPC 2.0 batch of ALL notifications...")
+        body = []
+        for cmd in commands:
+            body.append(builder.rpc_request(cmd, version=2, notification=True))
+        response, status = send_json_rpc(self.nodes[0], body)
+        assert_equal(status, 204) # HTTP_NO_CONTENT
+        assert_equal(response, None) # not even an empty array
+
     def test_http_status_codes(self):
         self.log.info("Testing HTTP status codes for JSON-RPC 1.1 requests...")
         # OK
```

Reply:
Pulled changes (771d1e1d206efe687b8661ab966cc1a62cc7ba39), built, ran all functional tests (all passed).  Re-ran the following on 771d1e1d206efe687b8661ab966cc1a62cc7ba39 and on release v26.0.


## Action (1): __Unexpected difference__
`curl --user test --data-binary '[]' -H 'content-type: text/plain;' http://127.0.0.1:38332/`

Result: Unexpected difference

v26.0 behavior: `[]`

PR 27101 (771d1e1d206efe687b8661ab966cc1a62cc7ba39) behavior: (no content) (in alignment with 2.0 spec, but not matching legacy behavior)
Applicable line(s):  httprpc.cpp:250
`if (reply.empty()) {` seems to be treating 2.0 and 1.x requests identically.
At first glance, seems like a check for 2.0 here might rememdy this (e.g., would need to be both 2.0 and reply.empty()).


## Action (2): __Minor Unexpected difference__
`curl --user test --data-binary '' -H 'content-type: text/plain;' http://127.0.0.1:38332/`

Result: Unexpected difference

v26.0 behavior: `{"result":null,"error":{"code":-32700,"message":"Parse error"},"id":null}`

PR 27101 (771d1e1d206efe687b8661ab966cc1a62cc7ba39) behavior: `{"result":null,"error":{"code":-32700,"message":"Parse error"}}` 
Diff: legacy includes `"id":null` while new code does not.


## Action (3): Expected difference
`curl --user test --data-binary '[{"jsonrpc":"2.0", "id":""}]' -H 'content-type: text/plain;' http://127.0.0.1:38332/`

Result: Expected difference

v26.0 behavior: `[{"result":null,"error":{"code":-32600,"message":"Missing method"},"id":""}]`

PR 27101 (771d1e1d206efe687b8661ab966cc1a62cc7ba39) behavior: `[{"jsonrpc":"2.0","error":{"code":-32600,"message":"Missing method"},"id":""}]`
Seems conformant to 2.0 spec.


## Action (4): Expected difference
`curl --user test --data-binary '[{"jsonrpc":"2.0"}]' -H 'content-type: text/plain;' http://127.0.0.1:38332/`

Result: Expected difference

v26.0 behavior: `[{"result":null,"error":{"code":-32600,"message":"Missing method"},"id":null}]`

PR 27101 (771d1e1d206efe687b8661ab966cc1a62cc7ba39) behavior: (no content) Seems compliant to 2.0 spec (batch with only a notification, albeit with a missing method, so no error response).

Reply:
re: https://github.com/bitcoin/bitcoin/pull/27101#discussion_r1473699881

> ## Action (1): **Unexpected difference**
> 
> `curl --user test --data-binary '[]' -H 'content-type: text/plain;' http://127.0.0.1:38332/`
> 
> Result: Unexpected difference
> 
> v26.0 behavior: `[]`
> 
> PR 27101 ([771d1e1](https://github.com/bitcoin/bitcoin/commit/771d1e1d206efe687b8661ab966cc1a62cc7ba39)) behavior: (no content) (in alignment with 2.0 spec, but not matching legacy behavior) Applicable line(s): httprpc.cpp:250 `if (reply.empty()) {` seems to be treating 2.0 and 1.x requests identically. At first glance, seems like a check for 2.0 here might rememdy this (e.g., would need to be both 2.0 and reply.empty()).

I think this is unavoidable if we want to comply with the 2.0 spec which says "If there are no Response objects contained within the Response array as it is to be sent to the client, the server MUST NOT return an empty Array and should return nothing at all."

Although from discussion thread above it sounds like maybe we are willing to not fully comply with 2.0 spec if we are prioritizing backwards compatibility over spec compatibility in cases where 2.0 spec is mandating a certain response even when the request does not contain a `"jsonrpc":"2.0"` field.

Either approach seems fine but we should probably note this corner case in the RPC doc if favoring compatibility over compliance with the spec, or note it in the release notes if we are going to favor compliance with the spec over compatibility in a case where jsonrpc 2.0 is not actually specified.

> ## Action (2): **Minor Unexpected difference**
> 
> `curl --user test --data-binary '' -H 'content-type: text/plain;' http://127.0.0.1:38332/`
> 
> Result: Unexpected difference
> 
> v26.0 behavior: `{"result":null,"error":{"code":-32700,"message":"Parse error"},"id":null}`
> 
> PR 27101 ([771d1e1](https://github.com/bitcoin/bitcoin/commit/771d1e1d206efe687b8661ab966cc1a62cc7ba39)) behavior: `{"result":null,"error":{"code":-32700,"message":"Parse error"}}` Diff: legacy includes `"id":null` while new code does not.

This is a little surprising. Seems better to keep previous behavior if this isn't actually specifying jsonrpc 2.0?

> ## Action (3): Expected difference

@tdb3, are actions 3-7 you listed bringing up potential bugs or they just recording what you tested? I didn't look closely all of these but on the surface they seem like expected behavior.

Reply:
>Although from discussion thread above it sounds like maybe we are willing to not fully comply with 2.0 spec if we are prioritizing backwards compatibility over spec compatibility in cases where 2.0 spec is mandating a certain response even when the request does not contain a "jsonrpc":"2.0" field.

Yeah, exactly this.  If/when support for jsonrpc 1.0 and 1.1 is removed, then the implementation could be updated to achieve full/closer compliance to the 2.0 spec.  If it's decided that jsonrpc 1.0/1.1 support is never to be removed, then we'd have to accept imperfect conformance to 2.0.

>we should probably note this corner case in the RPC doc if favoring compatibility over compliance with the spec, or note it in the release notes if we are going to favor compliance with the spec over compatibility in a case where jsonrpc 2.0 is not actually specified.

Agreed.  This would inform users (of 2.0) and increase awareness for any 1.0/1.1 removal PR.  @pinheadmz, how about something like this for doc/JSON-RPC-interface.md:

Existing text:
>A 2.0 request is identified by the presence of `"jsonrpc": "2.0"` in the request body. If that key + value is not present in a request, the legacy JSON-RPC v1.1 protocol is followed instead, which was the only available protocol in previous releases.

Addition:
```
Note: Since backwards compatibility is being prioritized at this time, 
and legacy JSON-RPC protocol is followed for responses to requests 
without presense of "jsonrpc": "2.0", some non-conformance to the 
JSON-RPC 2.0 specification will occur.  For example, a request 
containing an empty array ("[]", i.e. body containing an empty array) 
would receive a response of "[]" (empty array, matching legacy response 
behavior) while the JSON-RPC specification prohibits an empty array as 
a response to an all-notification batch.
```

>This is a little surprising. Seems better to keep previous behavior if this isn't actually specifying jsonrpc 2.0?

Agreed.  58cb22c83c1b72af7f3c580dc6261cf50e7ed2cb was observed to fix this, so now the response maches the legacy response.

>are actions 3-7 you listed bringing up potential bugs or they just recording what you tested? I didn't look closely all of these but on the surface they seem like expected behavior.

Yeah, thank you.  I should have been clearer.  Actions 3 through 7 are just recording what was tested.  All expected behavior (no bugs observed) for 3 through 7.